### DISCUSSION TITLE
Archive Picrin Scheme?

### DISCUSSION THREAD

**Discussion by omasanori** [picrin-scheme/picrin#356] [General]
The development has been stopped for five years, and I am thinking about clarifying the status by archiving the repositories.

Any opinions? @nyuichi @KeenS @koba-e964 @stibear @zeptometer @dcurrie @okuoku

(For ones not listed above: welcome! I have no intent to reject your opinion; it was compiled from the commit history.)

**Comment by lassik** [comment]
There seem to be some commits after the latest release. If the project is archived, would it be possible to make one final release?

We preserve old Scheme implementations under Scheme.org:

* Websites at https://conservatory.scheme.org/
* Git repos at https://gitea.scheme.org/conservatory
* Source and binary releases at https://files.scheme.org/

We'd be happy to add anything you can send from Picrin.

**Reply by omasanori** [reply]
Agreed.

**Comment by dcurrie** [comment]
I have no objection to archiving the repositories.

**Comment by KeenS** [comment]
I'm for archiving picrin

**Comment by zeptometer** [comment]
Thank you for bringing this topic up. I have no objection to archiving picrin.

**Comment by koba-e964** [comment]
+1 for archiving picrin

**Comment by omasanori** [comment]
OK. While I'd like to hear from others, especially @nyuichi (the original author), I will take some time for preparation in 2023. (Not a formal promise, but yeah, hopefully I will have some peaceful moment to work on this...)

- [x] Update README to describe the status quo
- [x] Tag a release as requested by @lassik
- [ ] Anything else?

**Reply by lassik** [reply]
Thanks!

Have any release tarballs of Picrin been made in the past? In the Git tags there's only `v0.1`.

**Reply by omasanori** [reply]
> Have any release tarballs of Picrin been made in the past? In the Git tags there's only v0.1.

I don't think so. Moreover, as you may know, tar and zip archives generated automatically by GitHub are not guaranteed to be stable in term of hash values ex. SHA-256. Thus, I will upload a tarball manually when I make a release, to make sure the hash value becomes stable.

**Reply by lassik** [reply]
I didn't know that - sounds pretty bad! Making your own tarball is a good idea.

Can you make a release candidate tarball first? I can test on all the BSDs that it builds successfully and the tests pass.

What's the relationship between https://github.com/picrin-scheme/libpicrin and this main picrin repo?

**Reply by omasanori** [reply]
It fails to build even on Linux today as-is, so we may need to fix the code first.

libpicrin is picrin's `lib/`. It was managed using `git subtree`.

**Reply by lassik** [reply]
(Instability of the hashes is explained here: https://github.blog/2023-02-21-update-on-the-future-stability-of-source-code-archives-and-hashes/)

**Comment by omasanori** [comment]
I had done `git bisect` to find the last commit that can be built successfully and tag it as `v0.2`. Something seems broken just after that commit.

**Comment by omasanori** [comment]
All issues and pull requests are closed as they will not be addressed. If @lassik agrees, I will close this discussion thread and archive this repository. The last call for requests!

**Comment by lassik** [comment]
Thanks for doing the bisect @omasanori! Good plan.

I am able to build the `v0.2` tag on MacOS. Most tests pass, but about 10 tests fail.

The build depends on `bin/picrin-bootstrap` which has been committed to git as a MacOS executable.

Buillding on Linux fails because Linux can't run MacOS executables. Perhaps there is a way around this.

Can you make `picrin-0.1.tar.gz` and `picrin-0.2.tar.gz` releases using the `git archive` command (and ideally, PGP sign them)? The commit history is complex enough that we should probably release these as-is with no changes. After all, the interpreter runs.

I can add the tarballs to https://files.scheme.org/?

**Reply by omasanori** [reply]
I uploaded OpenPGP signatures and the public key. See <https://github.com/picrin-scheme/picrin/releases>.

For additional record, I display the public key here:
```
-----BEGIN PGP PUBLIC KEY BLOCK-----
Comment: 1FA1 DE0E A52B 7AA1 B4DF  D563 7B43 95A4 0A28 DFC6
Comment: Masanori Ogino
Comment: mogino@acm.org

xjMEZeGojBYJKwYBBAHaRw8BAQdABBRIoMA0PFpajToT9SDsEPz7lHkrm3vNPGSv
HURriJrCwBEEHxYKAIMFgmXhqIwFiQWkj70DCwkHCRB7Q5WkCijfxkcUAAAAAAAe
ACBzYWx0QG5vdGF0aW9ucy5zZXF1b2lhLXBncC5vcmem/SagKewmOVCcxek+xM0Z
ohw7yHq0H/0UU2G/1bUzoQMVCggCmwECHgEWIQQfod4OpSt6obTf1WN7Q5WkCijf
xgAAKt0BAJleft68pcFTKkjbdd7ghOt6jcocggmX4DZrJXpyN7MpAQCWzQVsihjd
fW9F1Vf+JishbNk78339iCPzi7fZyEvmA80OTWFzYW5vcmkgT2dpbm/CwBQEExYK
AIYFgmXhqIwFiQWkj70DCwkHCRB7Q5WkCijfxkcUAAAAAAAeACBzYWx0QG5vdGF0
aW9ucy5zZXF1b2lhLXBncC5vcmfiXT3f0dVicm3rt1L+I7ey9kQOC/6/c0VJ6fz8
TcUcDAMVCggCmQECmwECHgEWIQQfod4OpSt6obTf1WN7Q5WkCijfxgAAIc4BAP1r
ujldaYS/rSqaj92C3SIdbCWt9DWrprTfqhYktv5MAP9mcTll5VncTh13cC2ffIJ6
EODoydUGum4IJf4DYEYnB80ObW9naW5vQGFjbS5vcmfCwBEEExYKAIMFgmXhqIwF
iQWkj70DCwkHCRB7Q5WkCijfxkcUAAAAAAAeACBzYWx0QG5vdGF0aW9ucy5zZXF1
b2lhLXBncC5vcmfv00GhF6VXm8vkylnHp4DhSn0uWBzQ/Wt1jE9+Ra/HUQMVCggC
mwECHgEWIQQfod4OpSt6obTf1WN7Q5WkCijfxgAAU8EA/3lmOXvTi+bBovTjRLUR
ToAMbk4fARQsKf0de/lQmdSWAP4r8u8zKxl41PIE4z6F0BrCgI2vG5x70WjtNiup
owEVAM4zBGXhqIwWCSsGAQQB2kcPAQEHQCYuUyW+5rhm+peqwf0bxSpcg4rWu96T
o+5NYme8ZDKawsDFBBgWCgE3BYJl4aiMBYkFpI+9CRB7Q5WkCijfxkcUAAAAAAAe
ACBzYWx0QG5vdGF0aW9ucy5zZXF1b2lhLXBncC5vcmd6BKEPqlcbTeXiSpE7CqoB
frVXCPH+mqO9v7N7JeInigKbIL6gBBkWCgBvBYJl4aiMCRBntm6yJiVhEEcUAAAA
AAAeACBzYWx0QG5vdGF0aW9ucy5zZXF1b2lhLXBncC5vcmfiypojdv/47g7NSIre
P3E7SRR/T/gj0+LSN4gUFcbjuhYhBAdXxYR1Ojkg6SZZIme2brImJWEQAABlZAD9
FsBLl4TNDPhaqOjqb3espApsD2qYeGzg0hvAGQOGXwABAJmUd7AlBo+3pIg2av6u
Mc24NF5rOjvvWjMnfGjmiIgBFiEEH6HeDqUreqG039Vje0OVpAoo38YAABtCAP9K
CaZzZVCufNmxS7swRH5jdTky7HU3jJh1yVp6tLtnVQEA+kid2+Yvpw+wURFjnA/E
xEFmyT6o0ZgPbuZ7QEv4cQvOMwRl4aiMFgkrBgEEAdpHDwEBB0BuxSN0qGMs1Lbz
JogWPIz+sVXuwXYEH6RGubvSeUBTbsLAxQQYFgoBNwWCZeGojAWJBaSPvQkQe0OV
pAoo38ZHFAAAAAAAHgAgc2FsdEBub3RhdGlvbnMuc2VxdW9pYS1wZ3Aub3JnNe10
8lUc1ohlYi1/a26L2iYvRdEJ8RnHyBr9tdzm2scCmwK+oAQZFgoAbwWCZeGojAkQ
XXP6XR/gsXtHFAAAAAAAHgAgc2FsdEBub3RhdGlvbnMuc2VxdW9pYS1wZ3Aub3Jn
bqWR/Wfx4j25hrpdge3Gi+Z5fOOFX7LEEuxDRR1JOq8WIQTiwATfpoLri5ehmCZd
c/pdH+CxewAAxwwA/2c8TQDxmXeJwGyAUxozvNJikeMyhs1CUvf4X+2EUFJrAPsF
oZ1wq6SClfLK2ZwvApRmMGQyMCguhxVIJHiljGbiCRYhBB+h3g6lK3qhtN/VY3tD
laQKKN/GAAAMFQEA8FAMSreUnV8hoYivQCQryGRCDjJ3tc1kRtB1k0Aekx0A+gMC
OqOfBz5pxnkDDDsQe24RXUA0egfggLfYhuK2BB0KzjgEZeGojBIKKwYBBAGXVQEF
AQEHQBk320ybZ+eiSegtGDbA4T+YodTV7OPYKdxmAXX0z6AGAwEIB8LABgQYFgoA
eAWCZeGojAWJBaSPvQkQe0OVpAoo38ZHFAAAAAAAHgAgc2FsdEBub3RhdGlvbnMu
c2VxdW9pYS1wZ3Aub3JnzYerZ6BWbYCwWuvn8XsppzJ98ZwsKBy9xd2sI9BLGYEC
mwwWIQQfod4OpSt6obTf1WN7Q5WkCijfxgAAqDgA/jNUJfAf14txjVYSgzG3F3wt
kMOaTm1zFkbHi13dNHMoAQCAp4hX0ES2MHP4GdZyvBPluaCiyGVgCrzLuMWy/7ge
Bg==
=yDx7
-----END PGP PUBLIC KEY BLOCK-----
```

> I can add the tarballs to https://files.scheme.org/?

Yes. It is FOSS, isn't it? :-)

**Reply by lassik** [reply]
Thanks! Both tarballs' metadata is now [committed to the `files.scheme.org` git repo](https://github.com/schemeorg-community/files.scheme.org/commit/def9be0713375550e2e70ff9a44f34c8b30982df) and tarballs are available for download from https://files.scheme.org/

**Comment by lassik** [comment]
If there is no opposition, I can also add Picrin to https://conservatory.scheme.org/ and https://gitea.scheme.org/conservatory. I am trying to slowly recruit a pool of schemers to do basic maintenance to old Scheme code there.

I could try to do some cleanup starting from top of the `v0.2` commit:

* Remove the dependency on `bin/picrin-bootstrap`, for example storing the C source code of `picrin-bootstrap` in git and building the executable from source.
* Make the failing tests pass.

**Reply by omasanori** [reply]
Yes please.

**Reply by lassik** [reply]
Did Picrin ever have a home page (other than a git browser)? If not, we could convert the `docs` directory to HTML and publish.

**Reply by omasanori** [reply]
> Did Picrin ever have a home page (other than a git browser)?

Nope AFAIK

**Reply by lassik** [reply]
HTML documentation now at https://conservatory.scheme.org/picrin/

**Comment by lassik** [comment]
Unless there is someone who understands the commits that came after `v0.2`, it might be best to save the `master` branch as `old-master` and cherry-pick from that branch in the future. The newer commits are so complex that we cannot hope to resolve them quickly.

**Reply by omasanori** [reply]
I chose `backup` as the name instead of `old-master`, feel free to make another branch if you like.

**Reply by lassik** [reply]
Since all the new commits  are by @nyuichi, perhaps the branch name could be `nyuichi`.

`git log --pretty='%an' v0.2..backup`

**Reply by lassik** [reply]
I renamed the backup branch to `nyuichi` on [my GitHub fork](https://github.com/lassik/picrin/tree/nyuichi) and the [Scheme Conservatory clone](https://gitea.scheme.org/conservatory/picrin/src/branch/nyuichi).

**Comment by lassik** [comment]
Re: bootstrap, it should be enough to rewrite `tools/mkboot.scm` in C. It's a very simple script that turns text files into C strings.

The relevant makefile fragment is:

```Makefile
lib/ext/boot.c: piclib/boot.scm piclib/library.scm
	cat piclib/boot.scm piclib/library.scm | bin/picrin-bootstrap tools/mkboot.scm > lib/ext/boot.c
```

If we replace `tools/mkboot.scm` with `tools/mkboot.c`, then we don't need `bin/picrin-bootstrap`.

**Reply by omasanori** [reply]
Yup rewriting it in C (or Perl, as it already depend on it and Perl won't vanish IHMO) would be nice.

**Reply by lassik** [reply]
Yes, Perl is a good idea. A Perl script doesn't need to be compiled first.

**Reply by lassik** [reply]
I can write one.

**Reply by lassik** [reply]
Impelmented at https://github.com/lassik/picrin/commit/78ad55ef16197d886da2b1d1a90fe69dd2e6caa0. OK?

**Reply by omasanori** [reply]
Merged, thanks!

**Reply by omasanori** [reply]
I cherry-picked some changes to the `master` branch, but b9cfbe827627c536b83f8715a24b0d8938ab1597 seems hard to pick without `mkboot.scm`... c3084963f71bee9ce682bbfa4f7466edac350ab4 has a bootstrap problem, too.

**Reply by lassik** [reply]
Does the new `master` build successfully? Is there an `.scm` file missing?

It seems the new experimental commits made big changes to implement pre-compiling. Big changes are likely to break stuff, and the benefit of pre-compiling is speed. Since the new work is not yet stable, Picin does not yet have existing users who depend on the extra speed.

If we merge experimental work, we are likely to complicate the build and unlikely to help people.

**Reply by omasanori** [reply]
> Does the new `master` build successfully?

Unless you won't try to regenerate `boot.c`, yes.

> Is there an `.scm` file missing?

Yes.

> It seems the new experimental commits made big changes to implement pre-compiling. Big changes are likely to break stuff, and the benefit of pre-compiling is speed. Since the new work is not yet stable, Picin does not yet have existing users who depend on the extra speed.
> 
> If we merge experimental work, we are likely to complicate the build and unlikely to help people.

Agreed.

**Reply by lassik** [reply]
Have you been able to reach @nyuichi? He may be the only one who can understand and fix the new code without spending a lot of time. If he doesn't have time to fix it, it's probably best that we keep the new work in his own branch.

**Reply by omasanori** [reply]
> If he doesn't have time to fix it, it's probably best that we keep the new work in his own branch.

Sure. Then I'd like to (finally) archive this repository, ack?

**Reply by lassik** [reply]
Acknowledged. It seems everyone is busy with other projects. If we save the branch, it can always be picked up later.

**Comment by omasanori** [comment]
OK, I will try. Thanks for your guidance, @lassik!

**Comment by lassik** [comment]
With `mkboot.pl`, the Linux build passes. [Log here](http://misc.lassi.io/2024/picrin-g311abc4d-docker-build.log.gz)

**Reply by lassik** [reply]
It seems the shared library is built using `-Werror`. There are many `implicit-fallthrough` warnings.

**Reply by omasanori** [reply]
Removed `-Werror`.

**Comment by lassik** [comment]
I just noticed that the documentation is also up on [readthedocs.io](https://picrin.readthedocs.io/en/latest/). It is sourced from this GitHub repo (`picrin-scheme/picrin`). I wonder who has access to maintain it there? Should we keep it up?

**Reply by omasanori** [reply]
I will check if I have permission later. Thanks for noticing.

**Comment by lassik** [comment]
Noticed that `mkboot.pl` (which I wrote) does almost the same job as the existing `mkloader.pl`. The latter is more idiomatic Perl code. I can send one more PR to refactor the two scripts.

**Reply by lassik** [reply]
https://github.com/picrin-scheme/picrin/pull/358

**Comment by omasanori** [comment]
I made two more releases and am going to archive the repository. Thanks all!
